web/themes/custom/usagov/usagov.theme

```diff
@@ -10,13 +10,158 @@ use Drupal\taxonomy\Entity\Term;
+  $node = \Drupal::routeMatch()->getParameter('node');
+  $variables['node'] = $node;
+
+  // If the FAQ checkbox is checked, then parse the content to find questions and answers
+  if ($node->hasField('field_faq_page') && $node->get('field_faq_page')[0]->value) {
+    $variables['faq'] = generate_faq($node);
+  }
 }
 
+/**
+ * generate_faq takes a node, analyses the body content,
+ * and returns an array of objects containing a question and answer
+ */
+function generate_faq(&$node) {
+  $faq = [];
+
+  // Set up a DomDocument with the node's body content
+  $node_body = $node->get('body')[0]->value;
+  $dom = new DomDocument();
+  $dom->loadHTML($node_body);
+  $elements = $dom->documentElement->childNodes[0]->childNodes;
+  $answer = new DomDocument();
+  $answer->loadHTML("<html><body></body></html>");
+
+  // Count the number of H2s in the body content
+  $numH2s = count($dom->getElementsByTagName('h2'));
+
+  if ($numH2s > 1) {
+    // There are multiple H2s
+    // Each H2 is a question
+    $question = "No Questions Found Yet";
+
+    //loop through elements of the body content
+    foreach ($elements as $element) {
+      // Check the type of element
+      if ($element->nodeName == "h2") {
+        // $element is a new question
+        // If there was an existing question, store it and the answer
+        if ($question != "No Questions Found Yet") {
+          $answer = faq_answer_inner_html($answer);
+          $faq[] = (object) ['question' => $question, 'answer' => $answer];
+        }
+
+        // Set up the new question
+        $question = $element->nodeValue;
+
+        // Set up the new answer
+        $answer = new DomDocument();
+        $answer->loadHTML("<html><body></body></html>");
+
+      }
+      elseif ($element->nodeName == "drupal-paragraph") {
+        // $element is an embedded paragraph
+        faq_process_paragraph($element, $answer);
+      }
+      elseif ($element->nodeName == "#text") {
+        // $element is a text node
+        // This is white space that we can ignore.
+      }
+      else {
+        // $element is some kind of content
+        // Append the content to the current answer
+        $importedNode = $answer->importNode($element, TRUE);
+        $answer->documentElement->childNodes[0]->appendChild($importedNode);
+      }
+    }
+
+    // If there was an existing question, store it and the answer
+    if ($question != "No Questions Found Yet") {
+      $answer = faq_answer_inner_html($answer);
+      $faq[] = (object) ['question' => $question, 'answer' => $answer];
+    }
+
+  }
+  else {
+    // There are not multiple H2s
+    // The H1 is the question
+    $question = $node->getTitle();
+
+    // Setup the answer
+    $answer->loadHTML("<html><body></body></html>");
+
+    // Append the intro to the answer
+    $introText = $node->get('field_page_intro')[0]->value;
+    $intro = $answer->createElement('p');
+    $intro->textContent = $introText;
```

Comment:
I decided to test what happens if I check "FAQ Page" on a page that doesn't make sense. On the home page, I get an error page and this in the log: 
```
NOTICE: PHP message: TypeError: Cannot assign null to property DOMNode::$textContent of type string in /var/www/web/themes/custom/usagov/usagov.theme on line 99
```

I think `$introText` is null here and  we need to check that it's valid before trying to use it. If it isn't, bail out (return null or an empty array) and set a message to the user that generating the FAQ failed. I think you'd do something like: 

```
\Drupal::messenger()->addError("Unable to create FAQ Page Schema for this page. Ask Chris what's up."); // That's a joke.
```

Reply:
I think the Page Intro field is just blank on the homepage so I simply added an if statement to skip adding the page intro to the answer if there is no intro text.